src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
> We need to have a deterministic means of constructor selection that is independent of DI container in any case, so I don't see how this is limiting or harmful in any way.

This I very strongly disagree. This is not compatible with supporting a different IGrainActivator (remember that we will ship in the box an alternate ServiceBasedGrainActivator that will not use use ActivatorUtilities, but will use the container to resolve the transient/scoped instances that a user can configure in their container, similar to [ServiceBasedControllerActivator](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Controllers/ServiceBasedControllerActivator.cs) ), but also disallow custom activators or 3rd party containers that will choose the constructor in the way they want.
Absolutely defining which constructor to use is an unnecessary constraint (and fundamentally wrong IMO). After all, why do we even support switching containers if we put even more constraints to the abstraction, especially where user code is involved.
As I mentioned, the metadata for the facet con be provided by other means (not just attributes in a parameter, or working completely outside of DI), without putting this big constraint in our DI system, for no good reason.

If you want a 100% bullet-proof approach that will work regardless of how users decide to activate the grains (without any container integration code, even if that would be trivial to provide), then a good approach is to do property injection, similar to how the ControllerContext is being injected into controllers (which in their API, there is no need for an arbitrary constructor selection, nor even a base Controller class at all, which I hope we can achieve at some point with not requiring a base Grain class too). You can see how this is implemented by looking [DefaultControllerPropertyActivator](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultControllerPropertyActivator.cs) and related classes, which runs after a controller instance was created in whichever way the user wants.

Also, looking at this decision from a different angle: Why put this restriction now and couple the constructor selection into GrainTypeData, when this adds absolutely zero value for it to be coupled so eagerly. You can pick this 1 constructor (and not make it configurable for now, even if some people don't like it) and cache it completely separate from GrainTypeData on first usage.

You mentioned you were going to write up an issue with an extended design. It would be nice to  see that, since this PR seems to be making a few assumptions that need to be validated and agreed first, and I'm not sure this thread is the right place for people to chime in.